### DISCUSSION TITLE
PWA, serviceWorker, and manifest

### DISCUSSION THREAD

**Discussion by MartyNZ** [oqtane/oqtane.framework#1474] [Q&A] [answered]
HI,
I'm starting to look into using Oqtane as a PWA. Is there any documentation on how to configure the manifest and serviceWorker files in an Oqtane tenant site?

**Comment by sbwalker** [comment]
Go to Site Settings in the UI, enable PWA and specify the icons. The service worker and manifest will generated automatically.

**Reply by MartyNZ** [reply]
Thanks @sbwalker, 
Already got that bit done. Now I want to edit the manifest and serviceWorker for the new site. How do I do that?

**Reply by sbwalker** [reply]
You cannot edit the manifest - it is not a file - it is generated by code. What do you want to change in the manifest? 

The serviceworker is a JavaScript file that is distributed in the wwwroot folder - but I am not sure why you would need to modify it? It is not designed to be edited as it will be overwritten on every upgrade.

**Reply by MartyNZ** [reply]
How does the manifest know what resources I want cached etc.?
How do I access functionality like push notification, location services, local storage, etc?

**Reply by sbwalker** [reply]
you can review the logic here: https://github.com/oqtane/oqtane.framework/blob/dev/Oqtane.Client/UI/ThemeBuilder.razor - note that it is complicated to inject JavaScript into a Blazor application as you need to use JSInterop - and most of the values are dynamically injected due to the multi-tenant nature of Oqtane. If there are additional properties that should be set in the manifest please provide the details.

**Reply by sbwalker** [reply]
The service worker is here: https://github.com/oqtane/oqtane.framework/blob/dev/Oqtane.Server/wwwroot/service-worker.js - note that it is not utilizing caching because offline support is not really a high priority scenario in Oqtane as most functionality requires server access - especially when running on Blazor Server.

**Reply by MartyNZ** [reply]
Hi @sbwalker ,
I can see the building of the manifest file in ThemeBuilder.razor as well as the serviceWorker.
My problem is the theme that the site uses doesn't contain the link to the manifest file. This is the same for the default Oqtane Theme and the Default Blazor Theme.
`
!-- stub the PWA manifest but defer the assignment of href --

  link id="app-manifest" rel="manifest" 

!--EndFragment--

Were either of these generated by ThemeBuilder.razor or do I create my own theme to implement the manifest file?

Re: serviceWorker. From my reading it appears that I should be able to store my sw.js file for each tenant in the relative site folder. ie. content>tenant>1>site>1 and that sw.js will have the scope of that folder/site and below. So a correctly referenced service worker in the 'start_url' field of the manifest file, generated by ThemeBuilder.razor should take care of it? 

Let me know if/where I've veered off course.

I understand your approach with Oqtane focusing on the server side. However, I am developing an online publishing platform that has to cater to an audience with internet access that can be spotty or none. So the "Cache/Update/Refresh" approach seems to be the best one for me. I'm relying on Oqtane's server speed to send updates _fast_ when users are connected.

Have a great weekend.

**Reply by sbwalker** [reply]
@MartyNZ Oqtane has a root razor page _Host.cshtml which is the entry point for the application and is not intended to be modified. It contains the reference to the PWA manifest you mentioned above: 

```
    <!-- stub the PWA manifest but defer the assignment of href -->
    <link id="app-manifest" rel="manifest" />
```
At runtime, the framework determines if a site is PWA enabled and if so, it assigns the href to a dynamically generated manifest. It also dynamically injects JavaScript into the page to register the service-worker.js which is located in the wwwroot folder. You currently cannot provide a different service-worker.js per site - it is designed to use the service-worker.js which is distributed with the framework. Making this extensible would require an enhancement to the framework to enable the specification of an alternate serviceworker.

**Comment by MartyNZ** [comment, marked as answer]
Hi @sbwalker,
So if I've followed you correctly. The only way to implement a PWA atm is to I use a 1 site = 1 Oqtane installation?

**Reply by sbwalker** [reply]
@MartyNZ If you have multiple sites in your installation you can enable the PWA option on each of them and a custom manifest will be dynamically generated for each one at run-time and sent to the browser (no actual file is created on disk - it is completely dynamic ). Each site will also have a link to the service-worker.js dynamically injected as well ( however as I mentioned above there is a common service-worker.js file which is used currently ). So the framework can definitely handle PWAs for multiple sites.

**Reply by MartyNZ** [reply]
Hi @sbwalker ,
As I'm sure you can tell by now, I'm new to all this so thanks for putting up with my questions.
And I have proven to myself that the browser has access to a manifest even though I can't see it or a link to it.

So in summation, the only 'constraint' is that every site within the Oqtane instance will be using the same serviceWorker.js as it's source. So all features implemented within that file will be available to all sites whether they want them or not. Have I got that correct? If so, no problem.

Again, Thanks for your time and just as an FYI: I started with Dotnet Nuke at version 4.4. I've always been what I call a 'website assembler', able to customise skins and 3rd party modules sufficiently to provide what the client was looking for. I've flirted with coding a couple of times but have finally recently made the decision to focus on mastering C# and Blazor. I'm a big fan of your work and when I discovered the Oqtane project and that it was being spearheaded by you I was in before I knew what it was!

I'm looking forward to getting to know this product and hopefully soon will have something more than moral support to contribute.
Between Oqtane and 2sxc from @iJungleboy, I'm looking forward to creating some interesting applications.

And now, I'm off to try and get my azure Oqtane site to work.

See you in another thread ;-)
